File: DataStructuresC++/Tree/LowestCommonAncestor.cpp
#include <stddef.h>
#include <iostream>

using namespace std;

class Node {
    public:
        int data;
        Node *left;
        Node *right;
        Node(int d) {
            data = d;
            left = NULL;
            right = NULL;
        }
};

class Solution {
    public:
  		Node* insert(Node* root, int data) {
            if(root == NULL) {
                return new Node(data);
            } else {
                Node* cur;
                if(data <= root->data) {
                    cur = insert(root->left, data);
                    root->left = cur;
                } else {
                    cur = insert(root->right, data);
                    root->right = cur;
               }

               return root;
           }
        }

/*The tree node has data, left child and right child 
class Node {
    int data;
    Node* left;
    Node* right;
};

*/
  
    Node *lca(Node *root, int v1,int v2) {
        while (true){
            int max = v1<v2?v2:v1;
            int min = v1<v2?v1:v2;
            if (max < root->data) {
                root = root->left;
            }
            else if(min > root->data){
                root = root->right;
            }
            else{
                break;
            }
        }
        return root;
    }

}; //End of Solution

File: DataStructures/Tree/LowestCommonAncestor.py


# Enter your code here. Read input from STDIN. Print output to STDOUT
'''
class Node:
      def __init__(self,info): 
          self.info = info  
          self.left = None  
          self.right = None 
           

       // this is a node of the tree , which contains info as data, left , right
'''

def lca(root, a, b):
  node = root
  while node:
    if max(a, b) < node.info:
      node = node.left
    elif min(a, b) > node.info:
      node = node.right
    else:
      break
  return node

